File: Services/internal/infrastructure/router.go
package infrastructure

import (
	"log"

	"github.com/ChangerzaryX1602/SkillSync/internal/handlers"
	"github.com/ChangerzaryX1602/SkillSync/pkg/auth"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/permission"
	"github.com/ChangerzaryX1602/SkillSync/pkg/role"
	"github.com/ChangerzaryX1602/SkillSync/pkg/role_permission"
	"github.com/ChangerzaryX1602/SkillSync/pkg/user"
	"github.com/ChangerzaryX1602/SkillSync/pkg/user_role"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"

	_ "github.com/ChangerzaryX1602/SkillSync/docs"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/swagger"
)

// SetupRoutes is the Router for GoFiber App
func (s *Server) SetupRoutes(app *fiber.App) {

	// Prepare a static middleware to serve the built React files.
	app.Static("/", "./web/build")

	app.Get("/swagger/*", swagger.HandlerDefault)

	// API routes group
	groupApiV1 := app.Group("/api/v:version?", handlers.ApiLimiter)
	{
		groupApiV1.Get("/", handlers.Index())
	}

	// App Repository
	userRepository := user.NewUserRepository(s.Resources)
	authRepository := auth.NewAuthRepository(s.Resources)
	permissionRepository := permission.NewPermissionRepository(s.Resources)
	roleRepository := role.NewRoleRepository(s.Resources)
	rolePermissionRepository := role_permission.NewRolePermissionRepository(s.Resources)
	userRoleRepository := user_role.NewUserRoleRepository(s.Resources)
	// auto migrate DB only on main process
	if !fiber.IsChild() {
		if err := s.MainDbConn.Exec("CREATE EXTENSION IF NOT EXISTS vector").Error; err != nil {
			log.Panicf("Failed to enable pgvector extension: %v", err)
		}
		if err := userRepository.Migrate(); err != nil {
			log.Panicf("Failed to migrate user repository: %v", err)
		}
		if err := models.MigrateTablePermissions(s.MainDbConn); err != nil {
			log.Panicf("Failed to migrate permissions: %v", err)
		}
		if err := utils.MigratePermission(s.MainDbConn); err != nil {
			log.Panicf("Failed to migrate permission data: %v", err)
		}
	}
	routerResources := handlers.NewRouterResources(s.JwtResources.JwtKeyfunc, s.MainDbConn)
	// App Services
	userService := user.NewUserService(userRepository)
	authService := auth.NewAuthService(authRepository, userService)
	permissionService := permission.NewPermissionService(permissionRepository)
	roleService := role.NewRoleService(roleRepository)
	rolePermissionService := role_permission.NewRolePermissionService(rolePermissionRepository)
	userRoleService := user_role.NewUserRoleService(userRoleRepository)

	// App Handlers
	user.NewUserHandler(groupApiV1.Group("/users"), routerResources, userService)
	auth.NewAuthHandler(groupApiV1.Group("/auth"), routerResources, authService)
	permission.NewPermissionHandler(groupApiV1.Group("/permissions"), routerResources, permissionService)
	role.NewRoleHandler(groupApiV1.Group("/roles"), routerResources, roleService)
	role_permission.NewRolePermissionHandler(groupApiV1.Group("/role/permissions"), routerResources, rolePermissionService)
	user_role.NewUserRoleHandler(groupApiV1.Group("/user/roles"), routerResources, userRoleService)
	// App Routes

	// Prepare a fallback route to always serve the 'index.html', had there not be any matching routes.
	app.Static("*", "./web/build/index.html")
}

File: hono-clean-architecture/internal/infrastructure/router.ts
import { Hono } from "hono";
import { Resources, RouterResources } from "../../pkg/models";
import { newIndexHandler } from "../handlers/handler_index";
import { getPool } from "../datasources/drizzle";
import { newTxManager } from "../../pkg/tx_context/tx_manager";
import { migratePermissions } from "../../pkg/utils/migrate-permission";

import { newUserRepository } from "../../pkg/user/repository";
import { newRoleRepository } from "../../pkg/role/repository";
import { newPermissionRepository } from "../../pkg/permission/repository";
import { newRolePermissionRepository } from "../../pkg/role_permission/repository";
import { newUserRoleRepository } from "../../pkg/user_role/repository";
import { newAuthRepository } from "../../pkg/auth/repository";

import { newUserService } from "../../pkg/user/service";
import { newRoleService } from "../../pkg/role/service";
import { newPermissionService } from "../../pkg/permission/service";
import { newRolePermissionService } from "../../pkg/role_permission/service";
import { newUserRoleService } from "../../pkg/user_role/service";
import { newAuthService } from "../../pkg/auth/service";

import { newUserHandler } from "../../pkg/user/handler";
import { newRoleHandler } from "../../pkg/role/handler";
import { newPermissionHandler } from "../../pkg/permission/handler";
import { newRolePermissionHandler } from "../../pkg/role_permission/handler";
import { newUserRoleHandler } from "../../pkg/user_role/handler";
import { newAuthHandler } from "../../pkg/auth/handler";

export async function setupRoutes(
  app: Hono,
  resources: Resources,
  routerResources: RouterResources
): Promise<void> {
  newIndexHandler(app);

  // Migrate default permissions and roles on startup
  try {
    await migratePermissions(resources.mainDbConn);
    console.log("✅ Permissions migrated successfully");
  } catch (err) {
    console.error("❌ Failed to migrate permissions:", err);
  }

  const apiV1 = new Hono();

  apiV1.get("/", (c) => {
    return c.json({ message: "SkillSync API v1", status: "ok" });
  });

  const txManager = newTxManager(getPool());

  const userRepository = newUserRepository(resources);
  const roleRepository = newRoleRepository(resources);
  const permissionRepository = newPermissionRepository(resources);
  const rolePermissionRepository = newRolePermissionRepository(resources);
  const userRoleRepository = newUserRoleRepository(resources);
  const authRepository = newAuthRepository(resources);

  const userService = newUserService(userRepository, roleRepository, userRoleRepository, txManager);
  const roleService = newRoleService(roleRepository);
  const permissionService = newPermissionService(permissionRepository);
  const rolePermissionService = newRolePermissionService(rolePermissionRepository);
  const userRoleService = newUserRoleService(userRoleRepository);
  const authService = newAuthService(authRepository, userService);

  newUserHandler(apiV1, routerResources, userService);
  newRoleHandler(apiV1, routerResources, roleService);
  newPermissionHandler(apiV1, routerResources, permissionService);
  newRolePermissionHandler(apiV1, routerResources, rolePermissionService);
  newUserRoleHandler(apiV1, routerResources, userRoleService);
  newAuthHandler(apiV1, authService);

  app.route("/api/v1", apiV1);
}
